[Remote Inspection] nth-child selectors may target the wrong element

whsieh opened the pull request:
#### aebf3a2151c5d375c404747ae5090e255984cae5
<pre>
[Remote Inspection] nth-child selectors may target the wrong element
<a href="https://bugs.webkit.org/show_bug.cgi?id=272612">https://bugs.webkit.org/show_bug.cgi?id=272612</a>
<a href="https://rdar.apple.com/126362427">rdar://126362427</a>

Reviewed by Megan Gardner and Abrar Protyasha.

Currently, the `N` used for `nth-child(N)` selectors when extracting targeted elements represents
the Nth index of that element type; instead, it should represent the Nth element child relative to
the parent (regardless of type). Fix this bug by adjusting `parentRelativeSelectorRecursive`, and
add an API test to cover these cases.

* Source/WebCore/page/ElementTargetingController.cpp:
(WebCore::findChild):
(WebCore::parentRelativeSelectorRecursive):
(WebCore::childIndexByType): Deleted.
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ElementTargetingTests.mm:
(TestWebKitAPI::TEST(ElementTargeting, ParentRelativeSelectors)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/element-targeting-5.html: Added.

Canonical link: <a href="https://commits.webkit.org/277449@main">https://commits.webkit.org/277449@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7ecd4f0d219896b8210959431e68f90e64be3a1c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47648 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26836 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50325 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50331 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43703 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49955 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32674 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24301 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/38791 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48230 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/24485 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/41064 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20088 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/21968 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/42257 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5697 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/43990 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/42671 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52223 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22684 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19010 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/46097 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23956 "Built successfully") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/45127 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24744 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6731 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23676 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/page/ElementTargetingController.cpp:
```diff
@@ -101,24 +101,28 @@ static inline bool querySelectorMatchesOneElement(Document& document, const Stri
-            result.index = result.childCountOfType;
-
-        result.childCountOfType++;
+            index = currentChildIndex;
```
Optional nit: an extra line here might help with readability.
```suggestion
            index = currentChildIndex;

```

whsieh committed:
[Remote Inspection] nth-child selectors may target the wrong element
https://bugs.webkit.org/show_bug.cgi?id=272612
rdar://126362427

Reviewed by Megan Gardner and Abrar Protyasha.

Currently, the `N` used for `nth-child(N)` selectors when extracting targeted elements represents
the Nth index of that element type; instead, it should represent the Nth element child relative to
the parent (regardless of type). Fix this bug by adjusting `parentRelativeSelectorRecursive`, and
add an API test to cover these cases.

* Source/WebCore/page/ElementTargetingController.cpp:
(WebCore::findChild):
(WebCore::parentRelativeSelectorRecursive):
(WebCore::childIndexByType): Deleted.
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ElementTargetingTests.mm:
(TestWebKitAPI::TEST(ElementTargeting, ParentRelativeSelectors)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/element-targeting-5.html: Added.

Canonical link: https://commits.webkit.org/277449@main
